TPC-DS workloads

cin opened the pull request:
This PR is to add support for TPC-DS data generation, query generation, and query execution workloads to spark-bench. Some cleanup is still needed (especially around the unit tests), but I'd like to get some feedback on the progress thus far. I haven't started any of the documentation yet. I'll start that after the code is closer to being finalized.

Commit:
initial commit for tpcds workload

Commit:
datagen nearly working

Commit:
first version that sorta works

Commit:
added config settings to spark-conf
fixed typo in tpc ds workload
added separator to show tables output

Commit:
working locally w/query selection too

Commit:
1-99 queries...not 0-99

Commit:
moving toward general data generation for tpc-ds

Commit:
latest before partial revert due to "TpcDsDataGen; no valid constructor" error

Commit:
revert back to how i want it to work. probably still have the "TpcDsDataGen; no valid constructor" error

Commit:
after looking farther down in the stack trace, the "TpcDsDataGen; no valid constructor" error being caused by some of the class member variables not being able to be serialized properly over to the executors for genData to run. had to turn those into broadcast variables and all was well.

Commit:
hopefully finally past the no valid constructor error. had to make TpcDsBase serializable.
asyncCopy seems to be working but needs more testing.
first working version with dsdgen..yay

Commit:
added some simple validation
took out the Broadcast variables as they weren't needed

Commit:
refactoring for readability

Commit:
WIP: in the middle of massive refactor to deal with table creation independently

Commit:
first step to getting tables built individually

Commit:
reordered imports

Commit:
Merge remote-tracking branch 'cin-bench/master' into tpcds-workload

Commit:
parallelized table generation
fixed issue with table option to dsdgen generating tables other than the table you specified
  parent tables produce their children as well
changed datagen workload to output timing information for each table's gen

Commit:
fixed bug with output location when generating data
refactored things a bit too. table-options is now just a seq instead of option seq. made this change bc the param is required to be present.
parallelized individual table creation

Commit:
updates and refactoring for unit tests

Commit:
TableOptions unit tests

Commit:
exposing private methods for testing
added unit tests for most methods. working on cleanup test currently

Commit:
added a couple more integration style tests

Commit:
refactored cleanup and journey existence methods and tests.
added genData tests

Commit:
fixed bugs in data generation and validation for tpcds.
unit/integration tests for tpcds are mostly complete. may add a doWorkload test

Commit:
removed unusesd import

Commit:
added table row count validation
added unit and integration tests for new code

Commit:
removed tpcds-journey stuff aside from ddl scripts
ddl scripts will probably be removed soon too as they don't support partitioning
fixed bug around tables that output children (*_returns tables)
included dsdgen and dsqgen binaries for unit/integration tests
  this may change in the future but needed a way to get tests to work with travis

Commit:
in process of adding support for query generation
refactor of general functions w/returns
added tests for general functions
added row  count to data gen stats

Commit:
apparently dsqgen doesn't deal well with absolute paths
thankfully scala's ProcessBuilder can take a current working directory argument

Commit:
refactored tpcds query workload to work without the journey's queries
started adding unit tests for tpcds query workload
removed unused import in tests

Commit:
changed query generation dialect from ansi to spark
found the spark.tpl in the journey
discovered this issue when testing the queries

Commit:
determined which queries were failing and why (for the most part)
added programmatic fix for query generation to deal with +/- x days syntax

Commit:
added more regexes to fix broken queries. hopefully have them all covered now.

ecurtin committed:
Adding JSON, Avro, and Orc formats for I/O

Commit:
added createtemptables config option to tpcds query workload
added baremetal configs to tpcds-example.conf

Commit:
removing most of the query templates from the tests.
probably broke some things but can't test right now due to other tests running on my machine

Commit:
fixed tpcds workload unit tests to match new test queries.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TableRowCounts.scala:
```diff
@@ -0,0 +1,78 @@
+
+object TableRowCounts {
+  // scalastyle:off magic.number
+  private[tpcds] val supportedScales = Seq(
```
you could make this a set

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TableRowCounts.scala:
```diff
@@ -0,0 +1,78 @@
+    100000  // 100TB
+  )
+
+  private val counts = Seq(
```
You could make this a map of table name -> scale -> count, and then the rest of the function in this object can go away.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    }
+  }
+
+  private def genData(implicit spark: SparkSession, exSvc: ExSvc): Seq[TpcDsTableGenStats] = {
```
this can have side effects so i would make it genData()(implicit ....)

cin replied:
good call

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    }
+  }
+
+  private def genData(topt: TableOptions, chile: Int) = {
```
What is a chile?

cin replied:
Hendrix! but sure that would be most ppl's reaction. will change.

brad-kaiser replied:
I thought it was like the country for kind of a long time.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    }
+  }
+
+  private[tpcds] def fixupOutputDirPath: String = output match {
```
output is a val so this will always have one value, so I would make this a val too not a def, and change the name to past tense, like fixedUpOutputDirPath.

cin replied:
good point.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    FileUtil.copy(file, dstFs, dstDir, false, conf)
+  }
+
+  private[tpcds] def asyncCopy(file: File, tableName: String)(implicit ec: ExSvc, conf: Configuration): Future[Boolean] = Future {
```
long line

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    syncCopy(file, s"$getOutputDataDir/$tableName")
+  }.recover { case e: Throwable => log.error(s"FileUtil.copy failed on ${file.getName}", e); false }
+
+  protected[tpcds] def createDatabase(implicit spark: SparkSession): Unit = {
```
this has side effects so it should have an empty parameter list

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    implicit val explicitlyDeclaredImplicitSpark = spark
+    implicit val ec = ExecutionContext.fromExecutorService(newFixedThreadPool(tableOptions.length))
+
+    val stats = try genData
```
the try and finally blocks should probably have {}s since catch has them

cin replied:
just looked this up in the databricks style guide and it enforces what you just said. changing...sadly i do this a lot.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGen.scala:
```diff
@@ -0,0 +1,274 @@
+    // TODO: optimization possible here.
+    // create the tables (convert from CSV to parquet) right after each table has been created
+    tables.foreach(createTable)
+    val rowCounts = validateRowCounts
```
it might be more clear if you got rid of the validateRowCounts function and just did "tables.map(validateRowCount)" here

cin replied:
agreed

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsQueryGen.scala:
```diff
@@ -0,0 +1,121 @@
+) extends Workload {
+  import TpcDsQueryGen._
+
+  private[tpcds] def mkCmd: Seq[String] = {
```
This can only have one value because its based on all vals, so make it a val too.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/ml/KMeansWorkload.scala:
```diff
@@ -133,7 +128,7 @@ case class KMeansWorkload(input: Option[String],
 
-  def save(ds: RDD[Vector], model: KMeansModel, spark: SparkSession): (Long, Unit) = {
+  def save(ds: RDD[Vector], model: KMeansModel, spark: SparkSession): Long = {
     val res = time {
```
I would do "val (time, result) = time {" so its easier to see whats going on with that tuple.

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/ml/KMeansWorkload.scala:
```diff
@@ -144,6 +139,6 @@ case class KMeansWorkload(input: Option[String],
     }
     ds.unpersist()
-    res
+    res._1
```
use the time val from last comment

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/tpcds/TpcDsWorkload.scala:
```diff
@@ -0,0 +1,129 @@
+  ) extends Workload {
+  import TpcDsWorkload._
+
+  private def getLines = Try {
```
this is a side effecting function so it should have parens

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/tpcds/TpcDsWorkload.scala:
```diff
@@ -0,0 +1,129 @@
+    val lines = getLines
+    val initialState = (QueryStateNone: QueryState, Option.empty[TpcDsQueryInfo], Seq.empty[TpcDsQueryInfo])
+    lines.foldLeft(initialState) {
+      case ((state, _, queryInfo), line) if state == QueryStateNone && line.startsWith("-- start query") =>
```
you can replace state and "if state == whatever" by just putting QueryStateNone in the case statement like
"case ((QueryStateNone, _, queryInfo), line) => ..."

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/tpcds/TpcDsWorkload.scala:
```diff
@@ -0,0 +1,129 @@
+  private[tpcds] def runQuery(queryInfo: TpcDsQueryInfo)(implicit spark: SparkSession): Seq[TpcDsQueryStats] = {
+    val queries = queryInfo.queries.mkString(" ").split(";").map(_.trim).filter(_.nonEmpty)
+    if (queries.isEmpty) throw new Exception(s"No queries to run for $queryStream")
+    log.error(s"Running TPC-DS Query ${queryInfo.queryTemplate}")
```
should this be logged as error? Shouldn't it be info or debug?

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/tpcds/TpcDsWorkload.scala:
```diff
@@ -0,0 +1,129 @@
+    log.error(s"Running TPC-DS Query ${queryInfo.queryTemplate}")
+
+    queries.map { query =>
+      val (dur, result) = time(spark.sql(query.replace(";", "")).collect)
```
might want to move the query.replace up with the rest of the query massaging at the beginning of the function

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/tpcds/TpcDsWorkload.scala:
```diff
@@ -0,0 +1,129 @@
+
+  override def doWorkload(df: Option[DataFrame], spark: SparkSession): DataFrame = {
+    implicit val explicitlyDeclaredImplicitSpark: SparkSession = spark
+    setup
```
has side effects, needs parens

brad-kaiser commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/tpcds/TpcDsWorkload.scala:
```diff
@@ -0,0 +1,129 @@
+  override def doWorkload(df: Option[DataFrame], spark: SparkSession): DataFrame = {
+    implicit val explicitlyDeclaredImplicitSpark: SparkSession = spark
+    setup
+    val queryStats = extractQueries.map { queryInfo => (queryInfo.queryNum, runQuery(queryInfo)) }
```
extractQueries has side effects too. needs parens.

brad-kaiser commented on examples/tpcds-example.conf:
```diff
@@ -0,0 +1,148 @@
+        benchmark-output = "console"
+        parallel = false
+        workloads = [
+//          {
```
can we get rid of commented lines

cin replied:
oops. those should actually be uncommented. was in a testing state where i didn't want to generated the data or queries.

brad-kaiser commented on utils/src/main/scala/com/ibm/sparktc/sparkbench/utils/GeneralFunctions.scala:
```diff
@@ -19,47 +19,30 @@ package com.ibm.sparktc.sparkbench.utils
-      case Some(a) => a
-    }
+  def getOrDefaultOpt[A](opt: Option[Any], default: A, func: Any => A = defaultFn[A](_: Any)): A = {
+    opt.foldLeft(default) { case (d, a) => tryWithDefault(a, d, func) }
```
i think opt.map(func).getOrElse(default) is clearer

cin replied:
good call!

brad-kaiser reviewed (approved):
LGTM. Left some comments. Thanks!

Commit:
added support for parallel query execution in tpcds query workload
updates based on Brad's code review comments

Commit:
changed tpcds query workload stats output
stats now flattened and include the query name, number, and index (for queries w/multiple phases)
fixed but introduced in code review changes (spark SQL doesn't like empty queries apparently)

showermat commented on cli/src/test/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGenTest.scala:
```diff
@@ -0,0 +1,482 @@
+  private val dbName = "testdb"
+
+  // TODO: is there a less brittle way to do this?
+  private val cwd = new java.io.File(".").getCanonicalPath
```
Does `sys.props("user.dir")` work?  If not, I feel like what you have is an acceptable workaround and you can remove the comment.

cin replied:
def like that better...was confusing that with `user.home`.

showermat commented on cli/src/test/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsDataGenTest.scala:
```diff
@@ -0,0 +1,482 @@
+    implicit val ec = ExecutionContext.fromExecutorService(newFixedThreadPool(1))
+    val f = workload.asyncCopy(tmpFile.toFile, "foo")
+    val results = waitForFutures(Seq(f))
+    results.forall(_ == true) shouldBe true
```
`results.foreach(_ shouldBe true)` or similar to eliminate the two `true`s?

showermat commented on cli/src/test/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsQueryGenTest.scala:
```diff
@@ -0,0 +1,136 @@
+class TpcDsQueryGenTest extends FlatSpec with Matchers {
+
+  // TODO: is there a less brittle way to do this?
+  private val cwd = new java.io.File(".").getCanonicalPath
```
Same as above.

showermat commented on utils/src/main/scala/com/ibm/sparktc/sparkbench/utils/GeneralFunctions.scala:
```diff
@@ -107,9 +92,28 @@ object GeneralFunctions {
 
+  // runCmd runs the command and checks the stderr to see if an error occurred
+  // normally the return value of the command would indicate a failure but that's not how tpc-ds rolls
+  def runCmd(cmd: Seq[String], cwd: Option[String] = None): Boolean = Try {
```
Is this useful for anything outside of TPC-DS?  If not, maybe it shouldn't be in `GeneralFunctions.scala`.

cin replied:
That's a fair point. I put it here in case future workloads needed it, but it could easily be moved to the TpcDsBase object.

showermat commented on cli/src/main/scala/com/ibm/sparktc/sparkbench/datageneration/tpcds/TpcDsQueryGen.scala:
```diff
@@ -0,0 +1,121 @@
+      lines.foreach { l => bw.write(l); bw.newLine() }
+      bw.flush()
+      bw.close()
+      Files.move(newf.toPath, ff.toPath, REPLACE_EXISTING)
```
Is it possible to overwrite the old file directly rather than moving?

Commit:
merged master

Commit:
not sure how this got back in  here...

Commit:
updates based on Matthew's comments.
also removed a return and println in SuiteKickoff

ecurtin committed:
Adding save-mode to workloads

ecurtin committed:
Adding error for wacky save-modes

ecurtin committed:
Disallow appending for workload results

Workload results are intended to be separate datasets. Allow appends
does not make sense conceptually.

ecurtin committed:
Documentation for save-mode additions

Commit:
added workaround to table-options
now being read from filesystem or jar

ecurtin committed:
Addressing code review comments

Moves format check to front.

Commit:
initial commit for partitioining support
added support for 10 and 100 GB dataset, just can't validate row counts

Commit:
fixed catalog_sales.sql
had partition statement in the wrong place
fixed partitioned query files

Commit:
made a couple log statement error
mainly so i don't have to mess with configs on kubernetes
added more spark-bench configs for data/query generation and query workloads against 1GB and 10GB datasets

Commit:
move tpcds examples to its own directory under examples
added a bunch of partitioned and nonpartitioned generation and query configs for 1 and 10 GB datasets

Commit:
added support for writing/ready query file generated by dsqgen from hdfs
could use some cleanup but focusing on functionality right now

Commit:
slight refactoring of output dir creation in query generator
added explain config flag to tpcds workload

Commit:
fixed file handle not getting closed properly in TpcDsBase.loadFile method
added unit tests for loadFile method
fixed a few unit test bugs that were introduced recently by moving stuff around
added test.db warehouse cleanup before tpcds data gen tests run (ensures clean slate in case tests were ctrl+c'd on a prior run)

ecurtin committed:
Bugfix for remote HDFS systems

Bugfix to allow users to provide hdfs hostname in their urls.

Commit:
i can't spel

Commit:
Merge branch 'origin-master' into tpcds-workload

# Conflicts:
#	cli/src/main/scala/com/ibm/sparktc/sparkbench/workload/Workload.scala
#	cli/src/test/scala/com/ibm/sparktc/sparkbench/workload/ml/KMeansWorkloadTest.scala

Commit:
updates to work with latest commits to master
probably broke some tpcds unit tests...

Commit:
added flag to enable hive support on spark session creation
added tpcds local configs for testing

Commit:
massive refactor to get things working with the spark on k8s fork
still undergoing testing

Commit:
updates to get spark-bench's TPC-DS workloads working with spark on kubernetes from the 2.3 branch

Commit:
configuration updates for latest round of TPC-DS benchmarks

Commit:
updates to TPC-DS benchmark configs so they can be executed interleaved.
added minimal confs to help test the interleaving script as well

cin commented:
closing this for now until validation and cleanup is done
